lib/graphql/connections/primary_key/base.rb

```
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+module GraphQL
+  module Connections
+    module PrimaryKey
+      # Base class for PrimaryKey pagination implementations
+      class Base < ::GraphQL::Connections::Base
+        COMPARABLE_METHODS = %i[
+          gt lt lteq gteq
+        ].freeze
+
+        def initialize(*args, primary_key: nil, **kwargs)
+          @primary_key = primary_key
+
+          super(*args, **kwargs)
+        end
+
+        def has_previous_page
+          if last
+            nodes.any? && items_exist?(type: :query, search: nodes.first[primary_key], page_type: :previous)
+          elsif after
+            items_exist?(type: :cursor, search: after_cursor, page_type: :previous)
+          else
+            false
+          end
+        end
+
+        def has_next_page
+          if first
+            items_exist?(type: :query, search: nodes.last[primary_key], page_type: :next)
+          elsif before
+            items_exist?(type: :cursor, search: before_cursor, page_type: :next)
+          else
+            false
+          end
+        end
+
+        def cursor_for(item)
+          cursor = serialize(item[primary_key])
+          cursor = encode(cursor) if opaque_cursor
+          cursor
+        end
+
+        private
+
+        def items_exist?(type:, search:, page_type:)
+          comparable_method = page_comparable_method(query_type: type, page_type: page_type)
+
+          return if COMPARABLE_METHODS.exclude?(comparable_method)
```

We shouldn't return nil from predicate methods. If we pass something illegal, we should raise the ArgumentError exception.

---

fixed